woocommerce/woocommerce-gateway-stripe
# Change display order upe methods

Fixes #2994

This PR implements the ability to order the payment methods on the Stripe settings page when UPE is enabled.

## Changes proposed in this Pull Request:
- Saving the UPE payment method order in an option `stripe_upe_payment_method_order`. If the option is empty but we have the order of legacy payment methods saved, we copy that order to this option to preserve the previous setting.
- Moved the capability check  to the backend.
- To render the payment methods in the saved order on the checkout page, we are reordering them in `reorder_available_payment_gateways` function similar to [the legacy one](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/4f2b0222221d1af815e991ee4f9d3518110d1c13/includes/class-wc-gateway-stripe.php#L658) 

## Testing instructions
- In `develop` branch enable the new checkout experience.
- On the Stripe settings page, note that there is no `Change display order` button at the top of the payment method list.
- Now enable the legacy checkout experience. You should see a `Change display order` button at the top of the payment method list.
- Change display order of a few methods and save changes.
- Now checkout to this (`task/2994-change-display-order-upe`) branch.
- Enable the new checkout experience and refresh the settings page.
- Confirm that the payment method order previously saved in legacy mode still persists and the additional UPE specific payment methods (BNPL, Wallet etc. which are not available in legacy mode) are placed at the end of the list.
- You should see a `Change display order` button at the top of the payment method list as well.
- Now change the display order of a few methods and save changes. If you refresh the page, the new order should persist.
- Enable a few methods randomly.
- As a shopper, add a product to the cart and go to the shortcode checkout page. Confirm that the available payment methods on the checkout page are following the order saved on the settings page.
- This PR does not cover the block checkout, so the order on the block checkout page will not follow the settings. It will be handled in a separate PR.
